src/contexts/WalletContext.tsx

```
@@ -0,0 +1,208 @@
+          side: type,
+          chain: chain,
+          wallet: wallet.getName().toLowerCase(),
+        },
+      });
+
+      const payload = {
+        address,
+        type: chainToPlatform(chain),
+        icon: wallet.getIcon(),
+        name: wallet.getName(),
+      };
+
+      if (type === TransferWallet.SENDING) {
+        dispatch(connectSourceWallet(payload));
+      } else {
+        dispatch(connectReceivingWallet(payload));
+      }
+
+      const handleAccountsChanged = (accounts: string[]) => {
+        console.log('Accounts changed:', accounts);
+        const shouldDisconnect =
+          accounts.length === 0 ||
+          (accounts.length && address && accounts[0] !== address);
+
+        if (shouldDisconnect) {
+          wallet.disconnect();
+        }
+      };
+
+      const handleDisconnect = () => {
+        console.log(`Wallet disconnected for ${chain} (${type})`);
+        dispatch(clearWallet(type));
+        wallet.off('disconnect', handleDisconnect);
+        wallet.off('accountsChanged', handleAccountsChanged);
+      };
+
+      wallet.on('disconnect', handleDisconnect);
+      wallet.on('accountsChanged', handleAccountsChanged);
+    },
+    [dispatch],
+  );
+
+  useEffect(() => {
+    provider.on('walletConnected', setupWalletConnection);
+
+    return () => {
+      provider.off('walletConnected', setupWalletConnection);
+    };
+  }, [provider, setupWalletConnection]);
+
+  const connectWallet = useCallback(
+    async (
+      chain: Chain,
+      type: TransferWallet,
+      connectLastUsed?: boolean,
+    ): Promise<Wallet> => {
+      setIsConnecting(true);
+      try {
+        let wallet: Wallet | null = null;
+        if (
+          connectLastUsed &&
+          isInternal &&
+          provider instanceof InternalWalletProvider
+        ) {
+          wallet = await provider.connectLastUsedWallet(chain, type);
+          if (!wallet)
+            throw new Error(`No last used wallet found for ${chain}`);
```

Is it expected the parent components catch this error? Alternative would be to have an error value in state, and return it as part of the context value, like an error code + message. Code for checking against.

BTW, Error() constructor can take a 2nd arg now [{ cause }](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause) which can be `{ code: string, error: Nested-Error-object }` or whatever shape you want.

---

Another question to the above, should we just do `else` part when !wallet above?

---

Connect did not previously have parent components catch any errors here, so I'm just maintaining parity for now by swallowing the errors. Maybe we can leave the behavior like this until we need the errors in parent components?